File: app/javascript/channels/scoring_channel.js
import consumer from "./consumer"

export function connectToScoringChannel(controller, userId, connectionCallback) {
  console.log("ScorringChannel: Connecting for user:", userId, "...");

  try {
    controller.scoringChannel = consumer.subscriptions.create(
      { channel: "ScoringChannel", user_id: userId },
      {
        connected() {
          console.log("ScoringChannel: Connected")
          connectionCallback(true)
        },

        disconnected() {
          console.log("ScoringChannel: Disconnected ")
          connectionCallback(false)
        },

        received(data) {
          console.info("ScoringChannel: Received data:", data);
          controller.handleWebSocketUpdate(data)
        }
      }
    )
  } catch (error) {
    console.error("ScoringChannel: Failed to create subscription", error);
    connectionCallback(false);
  }
}

File: app/channels/scoring_channel.rb
class ScoringChannel < ApplicationCable::Channel
  def subscribed
    # params[:user_id] is chosen by the client. Without this check any signed-in
    # team could subscribe to a rival's stream and watch their live scores --
    # including while the scoreboard is deliberately hidden. #receive was already
    # authorized; only reads were missed.
    user = User.find_by(id: params[:user_id])
    return reject if user.nil?
    return reject unless user == current_user || Ability.new(current_user).can?(:manage, :scoring)

    stream_for user
  end

  def receive(data)
    return unless Ability.new(current_user).can?(:update, Result)

    challenge_id = data["challenge_id"]
    user_id = data["user_id"]

    result = Result.find_or_initialize_by(challenge_id: challenge_id, user_id: user_id)
    result.regular_points = data["regular_points"]
    result.bonus_points = data["bonus_points"]
    result.updated_by_id = current_user.id

    if result.save
      Rails.logger.info "Updated score for user #{user_id}, challenge #{challenge_id}"
    else
      Rails.logger.error "Failed to update score: #{result.errors.full_messages.join(', ')}"
    end
  end
end
